src/serde/room_state.rs

```diff
@@ -0,0 +1,108 @@
+//! A module to deserialize a RoomState struct from incorrectly specified v1
+//! send_join endpoint.
+//!
+//! For more information, see this [GitHub issue](https://github.com/matrix-org/matrix-doc/issues/2541).
+
+use std::fmt;
+
+use serde::{
+    de::{self, Deserializer, Error, SeqAccess, Visitor},
+    ser::{SerializeSeq, Serializer},
+};
+
+use crate::v1::join::create_join_event::RoomState;
+
+pub fn serialize<S>(room_state: &RoomState, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let mut seq = serializer.serialize_seq(Some(2))?;
+    seq.serialize_element(&200)?;
+    seq.serialize_element(room_state)?;
+    seq.end()
+}
+
+pub fn deserialize<'de, D>(deserializer: D) -> Result<RoomState, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    deserializer.deserialize_seq(RoomStateVisitor)
+}
+
+struct RoomStateVisitor;
+
+impl<'de> Visitor<'de> for RoomStateVisitor {
+    type Value = RoomState;
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("Room State response")
+    }
+
+    fn visit_seq<V>(self, mut seq: V) -> Result<Self::Value, V::Error>
+    where
+        V: SeqAccess<'de>,
+    {
+        let bogus = seq
+            .next_element::<u32>()?
+            .ok_or_else(|| de::Error::custom("Expected integer as first element"))?;
+        if bogus != 200 {
+            return Err(V::Error::invalid_value(
+                de::Unexpected::Unsigned(bogus.into()),
+                &self,
+            ));
+        }
+
+        seq.next_element().map(|e| e.unwrap())
```

Comment:
I'm pretty sure `unwrap()`ping is wrong here. Will look at what `next_element` returns in detail tomorrow.

Reply:
I think this should be

```suggestion
        seq.next_element()?.ok_or_ese(|| V::Error::invalid_length(2, &"a second list element in the response"))
```

Reply:
Nice, makes sense. (I think after this endpoint I'll take some time to study effective handling of `Option`s and `Result`s.)